[RFR] Make a move

Luwangel opened the pull request:
- [x] Create methods to move the grid
- [x] Use the channel to create a game loop
- [x] Watch user inputs

Luwangel committed:
Create methods to make a move

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,146 @@
 type Grid [][]byte
 
+type Coords struct {
+	y int
```
`byte`

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,146 @@
+}
+
+func AreCoordsEquals(coords Coords, coords2 Coords) bool {
+	if coords.x == coords2.x && coords.y == coords2.y {
```
Can't `DeepEqual` be used here?

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,146 @@
+	return false
+}
+
+func IsGridResolved(grid Grid, startedGrid Grid) bool {
```
Duplicate of `AreGridsEquals` function

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,146 @@
 			value++
 			if value == size*size {
-				grid[i][j] = 0
+				grid[y][x] = 0
```
EMPTY_VALUE

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,146 @@
 	return grid
 }
+
+func DeepCopyGrid(grid Grid) Grid {
```
Try https://golang.org/pkg/reflect/#Copy?

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,146 @@
+	return false
+}
+
+func ValueFromCoords(grid Grid, coords Coords) byte {
```
Not sure we need a function for that. Or at least, ensure we don't go out of bounds.

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,146 @@
+
+func Move(grid Grid, coordsTileToMove Coords) (Grid, error) {
+	if !isTileInMovableTiles(grid, coordsTileToMove) {
+		return grid, errors.New("The tile is not movable")
```
Give coordinates in error message. It may help in the future and would be a good DX improvement.

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,146 @@
+	}
+
+	newGrid := DeepCopyGrid(grid)
+	newGrid[emptyCoords.y][emptyCoords.x] = grid[newCoords.y][newCoords.x]
```
To swap variables:

``` go
a, b = b, a
```

Luwangel committed:
Apply the review

Luwangel committed:
Create the user interface to move the puzzle

Luwangel committed:
Use the termbox terminal to move the puzzle with arrow keys

Kmaschta commented on src/game/game_test.go:
```diff
@@ -134,6 +134,40 @@ func TestListMovableTiles(t *testing.T) {
 	}
 }
 
+// Function CoordsFromDirection
```
Useless comments

Kmaschta commented on src/game/game_test.go:
```diff
@@ -134,6 +134,40 @@ func TestListMovableTiles(t *testing.T) {
+	}
+}
+
+// Function Move
```
Useless comments

Luwangel committed:
Remove comments from tests

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,179 @@
+
 const EMPTY_VALUE byte = 0
 
 func AreGridsEquals(grid Grid, grid2 Grid) bool {
```
Not sure we need another function for just one line.

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,179 @@
+	return reflect.DeepEqual(grid, grid2)
+}
+
+func AreCoordsEquals(coords Coords, coords2 Coords) bool {
```
Same

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,179 @@
+}
+
+func findEmptyTile(grid Grid) (Coords, error) {
+	return findTileByValue(grid, 0)
```
EMPTY_VALUE

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,179 @@
+
+	coordsEmptyTile, err := findEmptyTile(grid)
+	if err != nil {
+		return coordsMovableTiles
```
We should log error somewhere.

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,179 @@
+	}
+
+	size := byte(len(grid))
+	if coordsEmptyTile.y-1 != 255 {
```
I guess this is due to some binary shifting? I would instead use:

``` go
if coordsEmptyTile.y > 0
```

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,179 @@
+	if coordsEmptyTile.y+1 < size {
+		coordsMovableTiles = append(coordsMovableTiles, Coords{coordsEmptyTile.y + 1, coordsEmptyTile.x})
+	}
+	if coordsEmptyTile.x-1 != 255 {
```
Same

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,179 @@
+	}
+
+	size := byte(len(grid))
+	if dir == 's' || dir == 'S' {
```
Can't we use a `toLowerCase` like function?

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,179 @@
+
+func Move(grid Grid, coordsTileToMove Coords) (Grid, error) {
+	if !isTileInMovableTiles(grid, coordsTileToMove) {
+		return grid, errors.New(fmt.Sprintf("The tile ate coords (%d, %d) is not movable", coordsTileToMove.y, coordsTileToMove.y))
```
s/ate/at/

jpetitcolas commented on src/game/game.go:
```diff
@@ -1,38 +1,179 @@
+	return newGrid, nil
+}
+
+func MoveByValue(grid Grid, value byte) (Grid, error) {
```
You just use it in tests. Remove all dead code. We'll add it only when required.

jpetitcolas commented on src/main.go:
```diff
@@ -3,20 +3,123 @@ package main
+				}
+			case termbox.EventError:
+				panic(event.Err)
+			case termbox.EventInterrupt:
```
Isn't it a duplicate with `KeyCtrlC`?

jpetitcolas commented on src/main.go:
```diff
@@ -3,20 +3,123 @@ package main
+	clearCmd.Run()
+}
+
+func KeyListener(inputChan chan rune, doneChan chan bool) {
```
This file starts to grow too quickly. Can we split each go routine into a dedicated file?

Luwangel replied:
Is it possible to make that refactoring in another PR?

jpetitcolas commented on src/main.go:
```diff
@@ -3,20 +3,123 @@ package main
+	}
+	defer termbox.Close()
+
+	interruptChan := make(chan os.Signal, 2)
```
Do not forget to `defer close` all channels you open.

jpetitcolas reviewed (changes requested):
Great work so far! :)

sedy-bot committed:
Typo fix s/ate/at/

As requested by jpetitcolas at https://github.com/marmelab/15-puzzle-go/pull/3#discussion_r146262208

Luwangel committed:
Apply the review

jpetitcolas commented on src/main.go:
```diff
@@ -120,6 +118,12 @@ func main() {
 		fmt.Println("\nGGWP, you solved the puzzle!")
 	}
+
+	close(interruptChan)
```
You should `defer` the close. If an error occurs before reaching this point, your channel will never be closed.

Luwangel committed:
Defer the closing of channels

jpetitcolas commented on src/main.go:
```diff
@@ -119,10 +119,10 @@ func main() {
 		fmt.Println("\nGGWP, you solved the puzzle!")
 	}
 
-	close(interruptChan)
```
Move it just after the channel opening. Otherwise if an error occurs in `clearTerminal`, you will keep a channel opened even after the program close.

Luwangel committed:
Defer the closing at the beggining
